service/pkg/serviceregistry/serviceregistry.go

```diff
@@ -88,35 +79,129 @@ type DBRegister struct {
 // of the service within the instance of the platform.
 type Service struct {
 	Registration
+	impl       any
+	handleFunc HandlerServer
 	// Started is a flag that indicates whether the service has been started
 	Started bool
 	// Close is a function that can be called to close the service
 	Close func()
 }
 
-type ServiceMap map[string]Service
-type NamespaceMap map[string]ServiceMap
+// Start starts the service and performs necessary initialization steps.
+// It returns an error if the service is already started or if there is an issue running database migrations.
+func (s *Service) Start(ctx context.Context, params RegistrationParams) error {
+	if s.Started {
+		return fmt.Errorf("service already started")
+	}
+
+	if s.DB.Required && !params.DBClient.RanMigrations() && params.DBClient.MigrationsEnabled() {
+		appliedMigrations, err := params.DBClient.RunMigrations(ctx, s.DB.Migrations)
+		if err != nil {
+			return fmt.Errorf("issue running database migrations: %w", err)
+		}
+		params.Logger.Info("database migrations complete",
+			slog.Int("applied", appliedMigrations),
+		)
+	}
 
-// RegisteredServices is a map of namespaces to services
-// TODO remove the global variable and move towards a more functional approach
-var RegisteredServices NamespaceMap
+	s.impl, s.handleFunc = s.RegisterFunc(params)
 
-// RegisterService is a function that registers a service with the service registry.
-func RegisterService(r Registration) error {
-	if RegisteredServices == nil {
-		RegisteredServices = make(NamespaceMap, 0)
+	s.Started = true
+	return nil
+}
+
+// RegisterGRPCServer registers the gRPC server with the service implementation.
+// It checks if the service implementation is registered and then registers the service with the server.
+// It returns an error if the service implementation is not registered.
+func (s *Service) RegisterGRPCServer(server *grpc.Server) error {
+	if s.impl == nil {
+		return fmt.Errorf("service did not register an implementation")
 	}
-	if RegisteredServices[r.Namespace] == nil {
-		RegisteredServices[r.Namespace] = make(ServiceMap, 0)
+	server.RegisterService(s.ServiceDesc, s.impl)
+	return nil
+}
+
+// RegisterHTTPServer registers an HTTP server with the service.
+// It takes a context, a ServeMux, and an implementation function as parameters.
+// If the service did not register a handler, it returns an error.
+func (s *Service) RegisterHTTPServer(ctx context.Context, mux *runtime.ServeMux) error {
+	if s.handleFunc == nil {
+		return fmt.Errorf("service did not register a handler")
+	}
+	return s.handleFunc(ctx, mux, s.impl)
+}
```

Comment:
I think we should deprecate this. If we need we should require using gRPC and api gateway or connectrpc

Reply:
If we need it I'd rather make it difficult to use so that we encourage protos since they are selfdocumenting.